core/imageroot/var/lib/nethserver/cluster/actions/get-facts/50get

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2022 Nethesis S.r.l.
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+'''
+Get the facts for each of the agents below the cluster.
+'''
+
+import json
+import sys
+
+import agent
+import agent.tasks
+
+# setup redis connection
+redis_client = agent.redis_connect(privileged=True)
+
+# Init dictionary that returns the data
+facts = {
+    '$schema': 'https://schema.nethserver.org/facts/2022-12.json',
+    'uuid': '',
+    'installation': 'nethserver',
+    'facts': {
+        'cluster': [],
+        'nodes': {},
+        'modules': []
+    }
+}
+
+facts['uuid'] = redis_client.get('cluster/uuid')
+
+for key in redis_client.scan_iter('node/*/environment'):
+    # TODO: node_id not representative, change with NODE_UUID when available.
+    node_id = redis_client.hget(key, 'NODE_ID')
+
+    result = agent.tasks.run(
+        agent_id=f'node/{node_id}',
+        action='get-facts',
+        extra={
+            'isNotificationHidden': True,
+        },
+        endpoint="redis://cluster-leader"
+    )
+
+    # returns error if task failed
+    agent.assert_exp(result['exit_code'] == 0, result['error'])
```

Comment:
Are you sure to abort if just one node fails? Shall we continue with the next one?

Reply:
> Are you sure to abort if just one node fails?

In order to keep track of what happens to the cluster (more modules installed, node added, etc...) if a node fails, the phone-home will return that a node was removed. Since it's a thing that with the new installations appeared that we wanted to do, I've set a fail if the node fails to fetch info.

> Consider also to run all tasks in parallel with agent.runp_brief()

Already saw that, however only the error codes are returned with that function, but I need the whole output.

Reply:
If output is needed, you can still run in parallel with `agent.runp()`

The node task can fail because the node has been switched off. IIRC agent.run raises an exception in that case.

Reply:
> If output is needed, you can still run in parallel with agent.tasks.runp()

Missed the function 10 rows above :sweat_smile:

> The node task can fail because the node has been switched off. IIRC agent.run raises an exception in that case.

So anything that happens to agent run, if failed, throws an exception? In the phonehome bin script, if the get-facts fails no exception were thrown during my test. Maybe it depends on what happens inside the task? :man_shrugging:

Reply:
If the action fails an exit code is returned. If the task fails an exception is thrown.
Action failure is caused by a step. Task failure is caused by communication errors.